state-sync/inter-component/event-notifications/src/lib.rs

```diff
@@ -0,0 +1,447 @@
+    async fn notify_events(
+        &mut self,
+        version: Version,
+        events: Vec<ContractEvent>,
+    ) -> Result<(), Error> {
+        if events.is_empty() {
+            return Ok(()); // No events!
+        }
+
+        // Notify event subscribers and check if a reconfiguration event was processed
+        let reconfig_event_processed = self.notify_event_subscribers(version, events)?;
+
+        // If a reconfiguration event was found, also notify the reconfig subscribers
+        // of the new configuration values.
+        if reconfig_event_processed {
+            if self.reconfig_subscriptions.is_empty() {
+                return Ok(()); // No reconfiguration subscribers!
+            }
+
+            let new_configs = self.read_on_chain_configs(version)?;
+            for (_, reconfig_subscription) in self.reconfig_subscriptions.iter_mut() {
+                reconfig_subscription.notify_subscriber_of_configs(version, new_configs.clone())?;
+            }
+        }
+
+        Ok(())
+    }
+}
+
+/// A unique ID used to identify each subscription.
+type SubscriptionId = u64;
+
+/// A single event subscription, holding the subscription identifier, channel to
+/// send the corresponding notifications and a buffer to hold pending events.
+#[derive(Debug)]
+struct EventSubscription {
+    pub subscription_id: SubscriptionId,
+    pub event_buffer: Vec<ContractEvent>,
+    pub notification_sender: channel::diem_channel::Sender<(), EventNotification>,
+}
+
+impl EventSubscription {
+    fn buffer_event(&mut self, event: ContractEvent) {
+        self.event_buffer.push(event)
+    }
+
+    fn notify_subscriber_of_events(&mut self, version: Version) -> Result<(), Error> {
+        let event_notification = EventNotification {
+            subscribed_events: self.event_buffer.drain(..).collect(),
+            version,
+        };
+
+        self.notification_sender
+            .push((), event_notification)
+            .map_err(|error| Error::UnexpectedErrorEncountered(format!("{:?}", error)))
+    }
+}
+
+/// A single reconfig subscription, holding the channel to send the
+/// corresponding notifications.
+#[derive(Debug)]
+struct ReconfigSubscription {
+    pub subscription_id: SubscriptionId,
+    pub notification_sender: channel::diem_channel::Sender<(), ReconfigNotification>,
+}
+
+impl ReconfigSubscription {
+    fn notify_subscriber_of_configs(
+        &mut self,
+        version: Version,
+        on_chain_configs: OnChainConfigPayload,
+    ) -> Result<(), Error> {
+        let reconfig_notification = ReconfigNotification {
+            version,
+            on_chain_configs,
+        };
+
+        self.notification_sender
+            .push((), reconfig_notification)
+            .map_err(|error| Error::UnexpectedErrorEncountered(format!("{:?}", error)))
+    }
+}
+
+/// A notification for events.
+#[derive(Debug)]
+pub struct EventNotification {
+    pub version: Version,
```

Comment:
question: just to clarify, this is the "latest synced version" / block version and not like the transaction version right? was a bit confusing on my first read through : )